### PR TITLE
Feature/province list

### PR DESCRIPTION
**Pull request by EricDube**
# Description
# Dropdown
Closes: https://trello.com/c/KcuVOorR/181-people-can-type-to-select-their-province-or-territory-from-a-list
Allow users to select their province or territory from a drop down list. It allows the users to type in the province or territory and will auto-complete to the appropriate value. It does not allow values not in the list.

Screenshot from 2020-06-16 15-44-45

## Test Instructions

1. Navigate to the Your province or territory page
2. Enter text in text field or click on arrow of dropdown
3. Validate that the correct result displays on the results page.

# Back Button

Modified implementation of the Back button. The back button now stores a history of pages viewed in the session so the user can navigate back to the last page viewed with javascript disabled. 

Screenshot from 2020-06-16 16-06-49

## Test Instructions for back button changes

1. Navigate to the Your province or territory page
2. Select a province or territory and click Next
3. Validate that the back button appears at the top left of the current page
4. Click back button
5. Validate that it brings you back to the Your province or territory page
6. Validate that the back button does NOT appear on the Your province or territory page

## Help Requested
- Validate that the province drop down looks and behaves as expected.

### CODE REVIEW

**Comment by katedee** [comment]
One quick thing from the screenshot. I had a dark grey in the mocks, because that was just browser default for the list input. Could we keep the autocomplete default of black text on white? Or go with the dark grey as in designs? White text on black is a little hard to read. Thanks :)

**Comment by katewilhelm** [comment]
I've now finished testing that all provinces get the right result. Just the French ordering needs fixing as far as I can see

**Comment by CalvinRodo** [comment]
Province Sorting has been fixed, just need to fix unit tests

**Comment by katewilhelm** [comment]
Order of provinces looks great!

**Comment by CalvinRodo** [comment]
I've also addressed @katedee 's comments about the drop down colours

**Comment by dsamojlenko** [comment]
I'd recommend decoupling the js from the dropdown macro, and using the webpack build pipeline to compile the related js.

So:
- remove the `accessible-autocomplete.min.js` file from both `assets/js` and `public/dist/js`
- use npm to pull it into the project `npm install --save accessible-autocomplete`
- remove the js code from the `dropdown.njk` macro
- in `routes/question-province/js/question-province.js` import and setup the autocomplete:

```js
import accessibleAutocomplete from 'accessible-autocomplete'

accessibleAutocomplete.enhanceSelectElement({
  selectElement: document.querySelector('#province-select'),
  showAllValues: true,
  showNoOptionsFound: false,
})
```

- tell webpack about that file in `webpack.common.js`:

```js
entry: {
  styles: './assets/scss/app.scss',
  app: './assets/js/app.js',
  province: './routes/question-province/js/question-province.js',
},
```

- then include the compiled file in a script block in `routes/question-province/question-province.njk`:

```njk
{% block scripts %}
    <script type="text/javascript" src="{{ asset('/dist/js/province.js') }}"></script>
{% endblock %}
```

**Comment by corahansen** [comment]
Reviewed and approved, thank you!

File: app.js [hunk 1]
```diff
@@ -53,6 +53,10 @@ const app = express()
 app.set('trust proxy', 1)
 
 // general app configuration.
+app.use(function(req, res, next) {
+  res.set('Cache-Control', 'no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0');
+  next();
+});
```
Comment (CalvinRodo) [inline app.js:59]: This should go in the middleware folder

Comment (CalvinRodo) [reply app.js:59]: and have a test associated with it

File: app.js [hunk 2]
```diff
@@ -127,8 +131,20 @@ app.use(languageLinkHelper(app))
...
+  if (req.session === undefined || req.session.history === undefined || req.session.history.length === 0) {
+    res.locals.hideBackButton = true
+  } else {
+    res.locals.hideBackButton = false
+    req.session.history = req.session.history.filter(i => i === req.url)
+  }
+
+  next()
+})
+
```
Comment (CalvinRodo) [inline app.js:147]: This should go in the middleware folder and have a test associated with it

File: utils/history.helpers.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+  if (req.session.history === undefined) {
+    req.session.history = []
+  }
+  if (req.url !== undefined)
+  {
+    req.session.history.push(req.url)
+    // res.locals.hideBackButton = false
+  }
+
+}
```
Comment (CalvinRodo) [inline utils/history.helpers.js:12]: This should have a unit test associate with it.

**Pull request merged**